api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/computation/ExecutionTree.java

```diff
@@ -0,0 +1,95 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.computation;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.NullStep;
+
+public class ExecutionTree {
+    private Map<StepInfo, List<StepInfo>> forwardMap;
+    private Map<StepInfo, List<StepInfo>> backwardMap;
+    
+    private ExecutionTree() {
+        //map from previous to list of next
+        forwardMap = new HashMap<StepInfo, List<StepInfo>>();
+        //map from next to list of previous
+        backwardMap = new HashMap<StepInfo, List<StepInfo>>();
+    }
+    
+    public ExecutionTree(StepInfo rootNode) {
+        this();
+        addNodesToMaps(rootNode);
+    }
+
+    private void addNodesToMaps(StepInfo node) {
+        Set<StepInfo> nextNodes = node.getNextNodes();
+        for (StepInfo next : nextNodes) {
+            if (!containsPrev(next)) {
+                addArc(node, next);
+                addNodesToMaps(next);
+            }
+        }
+    }
+    
+    public List<StepInfo> getLeafs(){
+        return getPrevOf(NullStep.getInstance());
+    }
+
+    public List<StepInfo> getPrevOf(StepInfo node){
+        if (!backwardMap.containsKey(node)) {
+            return Collections.emptyList();
+        }
+        return backwardMap.get(node);
+    }
+    
+    public List<StepInfo> getNextOf(StepInfo node){
+        if (!forwardMap.containsKey(node)) {
+            return Collections.emptyList();
+        }
+        return forwardMap.get(node);
+    }
+    
+    private void addArc(StepInfo prev, StepInfo next) {
+        addToForwardMap(prev, next);
+        addToBackwardMap(prev, next);
+    }
+    
+    private boolean containsNext(StepInfo node) {
+        if (backwardMap.containsKey(node)) {
+            return true;
+        }
+        return false;
+    }
+    
+    private boolean containsPrev(StepInfo node) {
+        if (forwardMap.containsKey(node)) {
```

Comment:
Can be simply `return forwardMap.containsKey(node)`.

Reply:
Fixed